x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelAction.java

```diff
@@ -304,22 +308,106 @@ protected void masterOperation(
+        ActionListener<Void> storeModelListener
+    ) {
+        getExistingTaskInfo(client, modelId, isWaitForCompletion, ActionListener.wrap(taskInfo -> {
+            if (taskInfo != null) {
+                getModelInformation(client, modelId, sendResponseListener);
+            } else {
+                // no task exists so proceed with creating the model
+                storeModelListener.onResponse(null);
+            }
+        }, sendResponseListener::onFailure));
+    }
+
+    private static void getExistingTaskInfo(Client client, String modelId, boolean waitForCompletion, ActionListener<TaskInfo> listener) {
+        client.admin()
+            .cluster()
+            .prepareListTasks()
+            .setActions(MlTasks.MODEL_IMPORT_TASK_ACTION)
+            .setDetailed(true)
+            .setWaitForCompletion(waitForCompletion)
+            .setDescriptions("modelId=" + modelId)
+            .execute(ActionListener.wrap((response) -> {
+                var tasks = response.getTasks();
+
+                if (tasks.size() > 0) {
+                    // there really shouldn't be more than a single task but if there is we'll just use the first one
+                    listener.onResponse(tasks.get(0));
+                } else {
+                    listener.onResponse(null);
                 }
-                modelPackageConfigHolder.set(resolvedModelPackageConfig);
-                setTrainedModelConfigFieldsFromPackagedModel(trainedModelConfig, resolvedModelPackageConfig, xContentRegistry);
+            }, e -> {
+                listener.onFailure(
+                    new ElasticsearchStatusException(
+                        "Unable to retrieve task information for model id [{}]",
+                        RestStatus.INTERNAL_SERVER_ERROR,
+                        e,
+                        modelId
+                    )
+                );
+            }));
+    }
 
-                checkModelIdAgainstTags(trainedModelConfig.getModelId(), modelIdTagCheckListener);
-            }, listener::onFailure));
-        } else {
-            checkModelIdAgainstTags(config.getModelId(), modelIdTagCheckListener);
-        }
+    private static void getModelInformation(Client client, String modelId, ActionListener<Response> listener) {
+        client.execute(GetTrainedModelsAction.INSTANCE, new GetTrainedModelsAction.Request(modelId), ActionListener.wrap(models -> {
+            if (models.getResources().results().size() == 0) {
+                listener.onFailure(
+                    new ElasticsearchStatusException(
+                        "No model information found for a concurrent create model execution for model id [{}]",
+                        RestStatus.INTERNAL_SERVER_ERROR,
+                        modelId
+                    )
+                );
+            } else {
+                listener.onResponse(new PutTrainedModelAction.Response(models.getResources().results().get(0)));
```

Comment:
In the event that a task exists this will return the model information for the original request. So if the second request that hit this code path is attempting to change say the tags, it will be ignored. Do we think that's acceptable?

An alternative would be to return a failure saying that a download is occurring and it needs to be cancelled before we can attempt downloading again?

Reply:
Pre-packaged models don't allow changing the description, tags and other fields.

I wonder: can we distinguish between different downloads using the task API?

Reply:
> Pre-packaged models don't allow changing the description, tags and other fields.

Oh ok great.

> I wonder: can we distinguish between different downloads using the task API?

Yeah I believe that currently works, the logic I added only looks for tasks with a specific model id.